drivers/dai/intel/ssp/ssp.c

```diff
@@ -0,0 +1,2017 @@
+{
+	struct mn *mp = dai_get_mn(dp);
+	k_spinlock_key_t key;
+	uint32_t m = 1;
+	uint32_t n = 1;
+	int ret = 0;
+	bool mn_in_use;
+
+	key = k_spin_lock(&mp->lock);
+
+	mp->bclk_sources[dai_index] = MN_BCLK_SOURCE_NONE;
+
+	mn_in_use = dai_ssp_is_bclk_source_in_use(dp, MN_BCLK_SOURCE_MN);
+
+	if (dai_ssp_check_bclk_xtal_source(bclk_rate, mn_in_use, out_scr_div)) {
+		mp->bclk_sources[dai_index] = MN_BCLK_SOURCE_XTAL;
+		*out_need_ecs = false;
+		goto out;
+	}
+
+	*out_need_ecs = true;
+
+	if (mn_in_use)
+		ret = dai_ssp_setup_current_bclk_mn_source(dp, bclk_rate, out_scr_div, &m, &n);
+	else
+		ret = dai_ssp_setup_initial_bclk_mn_source(dp, bclk_rate, out_scr_div, &m, &n);
+
+	if (ret >= 0) {
+		mp->bclk_sources[dai_index] = MN_BCLK_SOURCE_MN;
+
+		LOG_INF("%s bclk_rate %d, *out_scr_div %d, m %d, n %d",
+			__func__, bclk_rate, *out_scr_div, m, n);
+
+		dai_mn_write(dp, MN_MDIV_M_VAL(dai_index), m);
+		dai_mn_write(dp, MN_MDIV_N_VAL(dai_index), n);
+	}
+
+out:
+
+	k_spin_unlock(&mp->lock, key);
+
+	return ret;
+}
+
+static void dai_ssp_mn_release_bclk(struct dai *dp, uint32_t dai_index)
+{
+	struct mn *mp = dai_get_mn(dp);
+	k_spinlock_key_t key;
+	bool mn_in_use;
+
+	key = k_spin_lock(&mp->lock);
+	mp->bclk_sources[dai_index] = MN_BCLK_SOURCE_NONE;
+
+	mn_in_use = dai_ssp_is_bclk_source_in_use(dp, MN_BCLK_SOURCE_MN);
+	/* release the M/N clock source if not used */
+	if (!mn_in_use)
+		dai_ssp_reset_bclk_mn_source(dp);
+	k_spin_unlock(&mp->lock, key);
+}
+
+static void dai_ssp_mn_reset_bclk_divider(struct dai *dp, uint32_t dai_index)
+{
+	struct mn *mp = dai_get_mn(dp);
+	k_spinlock_key_t key;
+
+	key = k_spin_lock(&mp->lock);
+
+	dai_mn_write(dp, MN_MDIV_M_VAL(dai_index), 1);
+	dai_mn_write(dp, MN_MDIV_N_VAL(dai_index), 1);
+
+	k_spin_unlock(&mp->lock, key);
+}
+
+static inline void dai_ssp_pm_runtime_dis_ssp_clk_gating(uint32_t index)
+{
+#if CONFIG_SOC_SERIES_INTEL_CAVS_V15
+	uint32_t shim_reg;
+
+	shim_reg = dai_shim_read(SHIM_CLKCTL) |
+		(index < DAI_NUM_SSP_BASE ?
+			SHIM_CLKCTL_I2SFDCGB(index) :
+			SHIM_CLKCTL_I2SEFDCGB(index - DAI_NUM_SSP_BASE));
+
+	dai_shim_write(SHIM_CLKCTL, shim_reg);
+
+	LOG_INF("%s index %d CLKCTL %08x", __func__, index, shim_reg);
+#endif
+}
+
+static inline void dai_ssp_pm_runtime_en_ssp_clk_gating(uint32_t index)
+{
+#if CONFIG_SOC_SERIES_INTEL_CAVS_V15
+	uint32_t shim_reg;
+
+	shim_reg = dai_shim_read(SHIM_CLKCTL) &
+		~(index < DAI_NUM_SSP_BASE ?
+			SHIM_CLKCTL_I2SFDCGB(index) :
+			SHIM_CLKCTL_I2SEFDCGB(index - DAI_NUM_SSP_BASE));
+
+	dai_shim_write(SHIM_CLKCTL, shim_reg);
+
+	LOG_INF("%s index %d CLKCTL %08x", __func__, index, shim_reg);
+#endif
+}
+
+static void dai_ssp_pm_runtime_en_ssp_power(uint32_t index)
+{
+#if CONFIG_SOC_SERIES_INTEL_CAVS_V25
+	uint32_t reg;
+
+	LOG_INF("%s en_ssp_power index %d", __func__, index);
+
+	sys_write32(sys_read32(I2SLCTL) | I2SLCTL_SPA(index), I2SLCTL);
+
+	/* Check if powered on. */
+	do {
+		reg = sys_read32(I2SLCTL);
+	} while (!(reg & I2SLCTL_CPA(index)));
```

Comment:
We need a timeout here so we never spin forever.

Reply:
ok